Implement a gRPC server which answer to discussed proto routes

remi-espie opened the pull request:
# Overview

This PR adds a gRPC server to the application, running it from main and answering to each discussed routes in pomegranate.proto.
However, nothing else is done as the Docker implementation is not finalized. The server answer the client with the given parameter.

# What's new?

- `server::start_server()` which start a gRPC server, and handle errors.

gRPC functions implemented:
- `start_deployment`, which will eventually launch a docker container. `project_uuid` is a string parameter.
- `restart_deployment`, which will stop and start a docker container. `project_uuid` is a string parameter.
- `delete_deployment`, which will delete a docker container. `project_uuid` is a string parameter.
- `stop_deployment`, which will stop a docker container. `project_uuid` is a string parameter.
- `deployment_status`, which will return the status of a docker container. `project_uuid` is a string parameter.
- `apply_config_deployment`, which will change the configuration of a docker container. `config` is a string parameter. Eventually a json string.

Note that every gRPC function do nothing and return the function name and its parameter as a string.

# Task list
- [X] Implement gRPC server
- [X] Implement every discussed gRPC routes
- [x] Give the docker engine as a parameter for the gRPC server *or make it a singleton ?*
- [x] Make gRPC functions do their docker works

# Other stuff

`tonic` and its dependencies (`prost` and `tokio`) were also added. 

~~A build dependency building the proto file with `tonic-build` was also added. 
For now, we are not using the [proto](https://github.com/paastech-cloud/proto) repo as a git submodule or a crate; things must be clarified with team.~~ Updated, see https://github.com/paastech-cloud/pomegranate/pull/4#issuecomment-1608946150

remi-espie committed:
feat: add pomegranate.proto following routes discussed in group
feat: add running gRPC server answering to pomegrenate.proto

remi-espie committed:
add protoc installation to build_and_test.yml ci

remi-espie committed:
fix(ci): remove protoc install from build_and_tests.yml and adds it to rust toolchain preparation

remi-espie committed:
fix: cargo format

Kuruyia commented on proto/pomegranate.proto:
```diff
@@ -0,0 +1,35 @@
+
+message ResponseMessage {
+  // Define your response message fields
+    string message = 1;
```
You can remove this message and fix indentation.
```suggestion
  string message = 1;
```

Kuruyia commented on src/main.rs:
```diff
@@ -33,4 +34,12 @@ fn main() {
 
     engine.start_application(&sample_app);
+
+    let _ = match server::start_server() {
+        Ok(server) => server,
+        Err(e) => {
+            info!("Failed to start gRPC server: {}", e);
+            return;
+        }
+    };
```
If you don't use the return value, can you return nothing on success and remove the assignation?

It would probably also be better to propagate the error and use a more meaningful log level.
```suggestion
    match server::start_server() {
        Ok(_) => Ok(()),
        Err(e) => {
            error!("Failed to start gRPC server: {}", e);
            Err(e)
        }
    }
```

mfernd commented on proto/pomegranate.proto:
Use the repo to save your proto file please... It's really simple...  
[paastech-cloud/proto](https://github.com/paastech-cloud/proto)

mfernd commented on build.rs:
You do not need this if you use [the proto repo](https://github.com/paastech-cloud/proto)

remi-espie committed:
fix(gRPC): adds deployment_status method

Kuruyia commented on src/server.rs:
```diff
@@ -0,0 +1,121 @@
+    }
+}
+
+#[tokio::main]
```
This should go on the `main()` function.

Kuruyia commented on src/server.rs:
```diff
@@ -0,0 +1,121 @@
+    async fn start_deployment(
+        &self,
+        request: Request<StartDeploymentRequest>,
+    ) -> Result<Response<ResponseMessage>, Status> {
```
The error will be a custom one after #2 is merged instead of `Status`.

Ditto for the other methods of this impl.

remi-espie committed:
fix(ci): remove dependencies from parallelizable jobs

remi-espie committed:
fix: typo

remi-espie committed:
fix: typo + indentation

remi-espie committed:
fix: more typo

remi-espie committed:
fix: remove unused unnamed variable
fix: catch the error better

remi-espie committed:
fix: remove unused unnamed variable
fix: catch the error better

remi-espie committed:
feat(proto): remove proto files and adds paastech-proto as dependency
(will need to update dependency when PR there is merged)

remi-espie commented:
[paastech-proto](https://github.com/paastech-cloud/proto) is now used as a dependency from a given commit ref. We will need to update that when PR there is merged.
Local proto files have then been deleted.

remi-espie committed:
fix(proto): remove build dependency

Kuruyia commented on src/server.rs:
```diff
@@ -0,0 +1,119 @@
+    }
+}
+
+pub(crate) async fn start_server() -> Result<(), Box<dyn std::error::Error>> {
```
No need to have a specific visibility scope imho.

```suggestion
pub async fn start_server() -> Result<(), Box<dyn std::error::Error>> {
```

Kuruyia commented on src/server.rs:
We might want to think about moving this file in a directory dedicated to gRPC if there's going to be more structs than `PomegranateGrpcService`.

remi-espie committed:
fix(grpc): rename server.rs to grpc_server.rs for understandability
fix(grpc): renamed a few variable to match their uses
fix(grpc): makes start_server function globally public
feat(proto): make proto dependency to main instead of given commit

remi-espie commented:
https://github.com/paastech-cloud/pomegranate/pull/4#issuecomment-1608946150 Updated to dependency to main branch !

remi-espie committed:
Merge branch 'main' into feat/grpc_server

# Conflicts:
#	Cargo.toml
#	src/main.rs

remi-espie committed:
feat(grpc): update codebase to work with DockerEngine

remi-espie committed:
feat(fmt): cargo fmt

remi-espie committed:
feat(dockerEngine): add DockerEngine implementation to gRPC routes
feat(grpc): adds fake db to retrieve data from and give it to DockerEngine
fix(typo): rename every deployment's `project_uuid` variable to `deployment_uuid`

remi-espie committed:
feat(docs): add documentation to gRPC server and its functions

remi-espie committed:
fix(fmt): format code

Kuruyia commented on src/grpc_server.rs:
```diff
@@ -0,0 +1,253 @@
+/// - The uuid of the application to get.
+/// # Returns
+/// The application.
+fn get_app(db: &Db, uuid: String) -> Application {
```
Maybe this should be in `Db`?

Kuruyia commented on src/db.rs:
```diff
@@ -0,0 +1,37 @@
+        Db {}
+    }
+
+    pub fn get_deployment_join_project(&self, uuid: String) -> DeploymentJoinProject {
```
I think it might be better to return a tuple of `(struct Deployment, struct Project)`.

That means that you have to create those two new structs.

Also, rename this function to `get_deployment_with_project`.

remi-espie replied:
Done differently but done nonetheless.

I chose to rename the function to `get_deployment_and_project`, which return a struct consisting of `Deployment` and `Project` (hence the name).

mfernd commented on src/grpc_server.rs:
```diff
@@ -0,0 +1,253 @@
+            .docker_engine
+            .get_application_status(app.project_id.as_str(), app.application_id.as_str())
+            .await
+            .unwrap();
```
:fearful:?

remi-espie replied:
🤔⁉️

mfernd replied:
`unwrap()` cause the program to panic (inside a thread I think it's just the thread that stops)

I don't see the use for it here

mfernd commented on src/main.rs:
```diff
@@ -22,4 +25,14 @@ async fn main() {
+
+    let db = db::Db::new();
+
+    match grpc_server::start_server(engine, db).await {
+        Ok(_) => Ok(()),
+        Err(e) => {
+            error!("Failed to start gRPC server: {}", e);
+            Err(e)
+        }
+    }
```
```suggestion
    grpc_server::start_server(engine, db).await.map_err(|e| {
        error!("Failed to start gRPC server: {}", e);
        e
    })
```

mfernd commented on src/grpc_server.rs:
```diff
@@ -0,0 +1,253 @@
+            env_variables: serde_json::from_str(&config).unwrap(),
+        };
+
+        let message: String = match self.docker_engine.restart_application(&app).await {
+            Ok(_) => match self.db.set_deployment_config(deployment_uuid, config).await {
+                Ok(_) => {
+                    format!("Applied config to application {}", app.project_id)
+                }
+                Err(e) => {
+                    format!(
+                        "Failed to save config to database {}: {}",
+                        app.project_id, e
+                    )
+                }
+            },
+            Err(e) => {
+                format!(
+                    "Failed to apply config to application {}: {}",
+                    app.project_id, e
+                )
+            }
+        };
```
It's strange to return every time an ok message and not using the `Err` with `Status`?

Something like this (example for failing to save config in db):

```rust
let msg = format!(
    "Failed to save config to database {}: {}",
    app.project_id, e
);
return Err(Status::new(Code::Internal, msg));
```

I have seen this here and in `start_deployment()`, `restart_deployment()`, `delete_deployment()`, `stop_deployment()`

remi-espie replied:
I didn't understood what you want; if the first match is ok then it does db save which in turn can return ok or err (?)

mfernd replied:
it's the goal of the `Result`, to return `Err` if there is an error, but here you're returning an error inside `Ok`

remi-espie committed:
feat(error): reworked error management to be more aligned with gRPC
fix(config): handle config parsing error
chore(refactor): moved get_app function to db.rs

remi-espie committed:
fix(fmt): cargo fmt

remi-espie committed:
feat(db): add struct for Deployment and Project independently and makes it transparent for grpc server

remi-espie committed:
fix(fmt): cargo fmt

Kuruyia commented on src/grpc_server.rs:
```diff
@@ -0,0 +1,287 @@
+            }
+        };
+
+        let response = ResponseMessage { message };
+        Ok(Response::new(response))
```
You should also send the status separately from the message to avoid a gRPC client from having to parse the message after calling this function.
The `.proto` needs to be changed.

remi-espie committed:
feat(grpc): add `status` string in gRPC `DeploymentStatusRequest`'s response to avoid parsing response message

remi-espie committed:
fix(fmt): cargo fmt

remi-espie committed:
fix(deployment status): replace simple struct to tuple

remi-espie committed:
fix(app): reworked app generation from their `uuid`

remi-espie committed:
feat(trace): add traces to every gRPC functions

remi-espie committed:
fix(docs): add documentation to `Project`, `Deployment` and `DeploymentAndProject`

remi-espie committed:
fix(fmt): cargo fmt

remi-espie committed:
feat(docs): specify returns values and errors for gRPC functions

remi-espie committed:
Merge branch 'main' into feat/grpc_server

# Conflicts:
#	Cargo.lock
#	Cargo.toml
#	src/main.rs

remi-espie committed:
feat(gRPC): modify scripts to remove database dependency
chore(db): remove db.rs

remi-espie committed:
feat(gRPC): add gRPC routes for logs and for stats

remi-espie committed:
feat(gRPC): delete application now delete image from registry

remi-espie committed:
chore(docs): update docs following documentation

remi-espie committed:
chore(readme): update README.md to specify to install protoc and where the proto files are located

Mathias-Boulay commented on src/grpc_server.rs:
```diff
@@ -0,0 +1,431 @@
+/// # Pomegranate gRPC server
+/// The gRPC server that implements the Pomegranate routes.
+pub struct PomegranateGrpcServer {
+    docker_engine: DockerEngine,
```
For the record, using a dyn Engine might be better if we want to abstract the engine used by the grpc server

Mathias-Boulay commented on src/grpc_server.rs:
```diff
@@ -0,0 +1,431 @@
+        let output = logs
+            .iter()
+            .map(|item| match item {
+                Ok(value) => from_utf8(value).unwrap().to_string(),
```
`.unwrap()` ?

remi-espie replied:
will change to `.unwrap_or_default()`

Mathias-Boulay commented on src/grpc_server.rs:
```diff
@@ -0,0 +1,431 @@
+/// # Returns
+/// Nothing.
+pub async fn start_server(docker_engine: DockerEngine) -> Result<(), Box<dyn std::error::Error>> {
+    let addr = "[::1]:50051".parse()?;
```
Shouldn't this be `[::]` to allow all addresses ?

Also, perhaps the port shouldn't be hardcoded

remi-espie replied:
Pomegranate is containerized, so exposing a hardcoded port seems better than having it set in a `.env` file, I believe ?

Mathias-Boulay reviewed (changes requested):
Only a handful of changes necessary

remi-espie committed:
fix(dockerfile): update Dockerfile to install protobuf-compiler during `builder` stage

remi-espie committed:
fix(dockerfile): expose grpc server port in Dockerfile and map it in docker-compose.yml

remi-espie committed:
fix(gRPC):
- replace `unwrap()` by `unwrap_or_default()`
- replace server answer address from `[::1]` to `[::]`

Mathias-Boulay commented on README.md:
````diff
@@ -33,3 +34,5 @@ cargo clippy
 cargo fmt --check
 ```
+
+Pomegranate's gRPC server will then run on `[::1]:50051`. The server will then answer to the proto routes defined at [paastech/proto](https://github.com/paastech-cloud/proto).
````
```suggestion
Pomegranate's gRPC server will then run on `[::]:50051`. The server will then answer to the proto routes defined at [paastech/proto](https://github.com/paastech-cloud/proto).
```

remi-espie committed:
fix: update README.md

Co-authored-by: Boulay Mathias <mathiasboulay@free.fr>

mfernd reviewed (approved):
:raised_hands:
